### PR TITLE
[Core][Chunked prefill] Skip sampling for chunked prefill and fix prompt logprob

### PR DESCRIPTION
**Pull request by rkooo567**
Prompt logprob is not working with chunked prefill becaues the existing logic has assumption the entire prefill is processed at once, and prompt log probs is calculated at once.

This assumption is incorrect after chunked prefill is enabled (we need to calculate prompt log prob for a chunk and incrementally update it).

To fix the issue, we introduce a new arg `do_sample` to SeqGruopMetadata. When a prefill is chunked, it is set to False, and when do_sample is False, we don't 1. calculate logprobs 2. do sampling. 

Question: I found not sampling chunked prefill doesn't make much performance diff. In this case, should we only not compute logprob instead of skipping the entire sampling which is a bit more complicated? 

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

File: tests/samplers/test_logprobs.py [hunk 1]
```diff
@@ -52,9 +70,18 @@ def test_get_prompt_logprobs(
             "The output text from the top logprob for each token position "
             "should be the same as the output text in the result.")
 
+        # The first prompt logprob is always None
```
Comment (rkooo567) [inline tests/samplers/test_logprobs.py:73]: NOTE: The current vLLM API returns None for the first index prompt logprob. 

So promt_logprob[0] -> the logprob the first token is likely to be generated given the 0th toekn, which doesn't exist. So it is always None.
prompt_logprob[1] -> the logprob the second prompt token is likely to be generated given the first token.

File: vllm/model_executor/layers/sampler.py [hunk 2]
```diff
@@ -120,7 +123,7 @@ def _apply_min_tokens_penalty(
         if (i < sampling_metadata.num_prompts
                 and sampling_params.prompt_logprobs is not None):
             assert len(seq_ids) == 1
-            start_idx += sampling_metadata.prompt_lens[i] - 1
+            start_idx += sampling_metadata.subquery_lens[i] - 1
```
Comment (rkooo567) [inline vllm/model_executor/layers/sampler.py:126]: we should use subquery len here. Prompt_len contains entire prompt token's length

File: vllm/model_executor/layers/sampler.py [hunk 3]
```diff
@@ -247,24 +255,30 @@ def _random_sample(
...
     is_prompts: List[bool],
     random_samples: torch.Tensor,
+    do_samples: List[bool],
 ) -> List[Tuple[List[int], List[int]]]:
     # Find the maximum best_of value of the prompt phase requests.
     random_samples = random_samples.cpu()
     sample_idx = 0
     results = []
-    for seq_group, is_prompt in zip(selected_seq_groups, is_prompts):
+    for seq_group, is_prompt, do_sample in zip(selected_seq_groups, is_prompts,
```
Comment (rkooo567) [inline vllm/model_executor/layers/sampler.py:264]: most of this part is just indented (no logic change except that if do_sample is False, it doesn't do anything)

File: vllm/model_executor/layers/sampler.py [hunk 4]
```diff
@@ -287,36 +302,42 @@ def _beam_search_sample(
...
-        if is_prompt:
-            # Prompt phase.
-            assert num_parent_seqs == 1, (
-                "Prompt input should have only one seq.")
-            parent_ids = [0] * (2 * beam_width)
-            _, next_token_ids = torch.topk(seq_group_logprobs[0],
-                                           2 * beam_width)
-            next_token_ids = next_token_ids.tolist()
+        if do_sample:
+            beam_width = sampling_params.best_of
```
Comment (rkooo567) [inline vllm/model_executor/layers/sampler.py:310]: just indented

File: vllm/model_executor/layers/sampler.py [hunk 5]
```diff
@@ -546,56 +586,97 @@ def _get_logprobs(
...
+
+    Returns:
+        A tuple of prompt and sample logprobs per sequence group in a batch.
+        If a seq_group doesn't require sample (e.g.,
+        sampling_metadata.do_samples[i] is False), the sampled logprobs can be
+        an empty string. If a seq group doesn't require prompt logprobs (e.g.,
+        if prompt_logprobs is None), the prompt logprobs may be None.
+    """
+    # The index of query token we care.
+    query_indices: List[int] = []
```
Comment (rkooo567) [inline vllm/model_executor/layers/sampler.py:610]: Most of code change here is the rename of existing variables. Main logic is when do_sample is False, it doesn't compute logprobs

**Comment by zifeitong** [comment]
Thanks, I tried this PR with `lm_eval_harness` and it worked well.

It also fixes another OOM problem I keep seeing. Previously, [Sampler.forward() ](https://github.com/vllm-project/vllm/blob/a53222544c6385ee314a26fdf42eb14f5b4e5ad9/vllm/model_executor/layers/sampler.py#L38) requires `O(N_prompt_lens * N_vocab_size)` of memory, when `prompt_logprobs` is requested. With chucked prefill, it's reduced to `O(subquery_lens * N_vocab_size)`

**Pull request closed without merging**